File: detectors/lengths/detectLengths.py
import os
import sys
import math
import random
import hashlib

sys.path.append(os.path.abspath('../..'))

from util import load

def mean(a):
  return float(sum(a))/float(len(a))

def dev(a, b):
  total=0
  for x in range(1500):
    diff=abs(a[x]-b[x])
    total=total+diff
  return total

def count(a, k):
  histogram = [1] * k
  for x in a:
    x=x[0]
    histogram[x-1] += 1
  return histogram

def mkdata(lengths):
  k = max(1500, max(map(len, lengths)))
  data=count(lengths, k)

  return data

labels=['obfsproxy', 'SSL', 'Dust']

res1=load('obfsproxy.csv', int)
res2=load('SSL.csv', int)
res3=load('Dust.csv', int)

files=load('files.csv')

d1s=[]
d2s=[]
decisions=[]

f=open('output.csv', 'wb')
for file in files:
  file=file[0]
  correct=file.split('/')[-2]
  if correct=='http':
    continue
  print('correct: '+str(correct))
  lengths=load(file, int)
  data=mkdata(lengths)

#  res1=[res1[0]]
#  res2=[res2[0]]
#  res3=[res3[0]]

  devs=[]
  for i in range(len(res1)):
    devs.append(dev(data, res1[i]))
  d1=math.log(mean(devs))

  devs2=[]
  for i in range(len(res2)):
    devs2.append(dev(data, res2[i]))
  d2=math.log(mean(devs2))

  devs3=[]
  for i in range(len(res3)):
    devs3.append(dev(data, res3[i]))
  d3=math.log(mean(devs3))

  ds=[d1,d2,d3]
  md=min(ds)
  di=ds.index(md)
  decision=labels[di]
  print(decision)

  f.write('lengths,'+str(file)+','+str(d1)+','+str(d2)+','+str(d3)+','+str(decision)+','+str(correct)+','+str(int(decision==correct))+"\n")
f.close()

File: process/src/detectors/lengths/detectLengths.r
library(arm)
library(coda)
library('plyr')

dev=function(a, b) { sum((a-b)^2) }
mkdata=function(lengths)
{
  K = max(c(1500, max(lengths$V1)))
  data=rep(1,K)
  freq=count(lengths$V1)
  for(i in seq(1,nrow(freq)))
  {
    data[freq[i,1]]=freq[i,2]+1
  }

  data
}

res1 = read.coda("obfsproxy/CODAchain1.txt", "obfsproxy/CODAindex.txt")
res2 = read.coda("SSL/CODAchain1.txt", "SSL/CODAindex.txt")

files=read.csv('files.csv', header=FALSE)
files=files$V1
d1s=c()
d2s=c()
decisions=c()

for(file in files)
{
  print(file)
  lengths=read.csv(file, header=FALSE)
  data=mkdata(lengths)

  devs=c()
  for(i in seq(1,nrow(res1)))
  {
    devs=c(devs, dev(data, res1[i,]))
  }
  d1=log(mean(devs))
  d1s=c(d1s, d1)

  devs2=c()
  for(i in seq(1,nrow(res2)))
  {
    devs2=c(devs2, dev(data, res2[i,]))
  }
  d2=log(mean(devs2))
  d2s=c(d2s, d2)

  show(c(d1, d2))

  if(d1<d2)
  {
    print('obfsproxy')
    decision='obfsproxy'
  }
  else if(d1>d2)
  {
    print('SSL')
    decision='SSL'
  }
  else
  {
    print('Undecided')
    decision='Undecided'
  }

  decisions=c(decisions, decision)
}

frame=data.frame(file=files, d1=d1s, d2=d2s, decision=decisions)
write.csv(frame, file='output.csv')

